### PR TITLE
Quick favorite and pin actions in list screen

### LINKED ISSUE #164: Add "pinned" and "quick favorite" features in list screen

### PR DESCRIPTION
**Pull request by AlexandreBeeDeeBop**
Content of this PR :
The user can swipe on a card to add it quickly as favorite or pin it above all other cards

How and Why :
By swiping left, the user can add quickly a parking spot to his favorites. Had there not been this feature, the user would have to click on the card and add it to his favorites from the parking detail screen, which takes more time and might make him lose track of where he was in the list. An unsigned user can still see this feature but will only see on swipe a message advertising the benefits of signing in, like having your own favorite parking spots.
By swiping right, the user can pin a parking spot. This will launch the card on top of the list in a different section.
Of course, there is also the option to unpin an already pinned card. This pin feature can help a user to come back to a parking spot later without having to look back for it in the big messy list later on, and also to not have to necessarily add it to favorites if he's interested but not that interested.
~~The pin is non-persistent, so if you switch screens it will unpin. If the team wants it to become persistent, this would have to wait for another sprint or two because it requires changes in the userViewModel in my opinion.~~
Actually, the pin is now persistent. It was the result of the necessity to fix the pin's behavior.

TestReport : 
image

Closes #164

### CODE REVIEW

**Review by Armetiis — COMMENTED** [review]
So, after testing we agreed that there was a very unexpected behaviour with the pins. After around 2 hours of debugging, I FINALLY found where it came from. It was related to the indexing of the LazyColumn (lazycolumn needed a key argument)

Here is a patch that solves this issue. Note that this also makes the pins persistant, but I think it's a positive thing. Please apply this patch. I will review once this is done
[fix-pin-unexpected-behaviour.patch](https://github.com/user-attachments/files/17747982/fix-pin-unexpected-behaviour.patch)

**Comment by AlexandreBeeDeeBop** [comment]
> So, after testing we agreed that there was a very unexpected behaviour with the pins. After around 2 hours of debugging, I FINALLY found where it came from. It was related to the indexing of the LazyColumn (lazycolumn needed a key argument)
> 
> Here is a patch that solves this issue. Note that this also makes the pins persistant, but I think it's a positive thing. Please apply this patch. I will review once this is done and you rerequest me [fix-pin-unexpected-behaviour.patch](https://github.com/user-attachments/files/17747982/fix-pin-unexpected-behaviour.patch)

🐐

**Comment by AlexandreBeeDeeBop** [comment]
I dont know about you guys but I HATE the new colours of the screen. The white background on white cards it just awful and now it's even harder to tell if we're in the pinned list or the normal list. Should I change it back to the sort of blue I had before the rebase ?

**Comment by Armetiis** [comment]
> I dont know about you guys but I HATE the new colours of the screen. The white background on white cards it just awful and now it's even harder to tell if we're in the pinned list or the normal list. Should I change it back to the sort of blue I had before the rebase ?

I think white cards on white background are alright. What we can do (but I think this is more of a concern for a future sprint) is remove lateral paddings and remove the rounded edges

File: app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt
Comment (Armetiis) [inline app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt:417]: Please can you remove the color argument that André's PR carefully removed? Also valid for line 345

File: app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt
Comment (Armetiis) [inline app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt:436]: Here too, make sure to remove the color argument that André carefully removed

**Review by LouisJamesVasseur — APPROVED** [review]
UI testing looks good, and so does the code. Good job!

**Review by Armetiis — CHANGES_REQUESTED** [review]
##### General
Nice PR. It adds a lot of dynamism to the app and enhances the accessibility. Good job!

However, the PR is lacking a video to showcase the new feature and this is a shame, because it's a really visual one.
Make sure to provide a coverage screenshot

##### Now regarding the code:
Rather than leaving you a comment for each line you didn't use string resources, I think it's better if I centralize them all here: 132, 153, 337 (twice), 347, 353, 381, 385

I also left you some comments in the code regarding some UI changes that you overwrote with respect to André's theme PR. Please make sure to checkout on main and compare everything. Now regarding functionality, everything seems good on my side

##### Remarks about the functionality

Note that the following remarks are non blocking and can be addressed in a future sprint.

Don't you think it would make sense to display all the pinned parkings, rather than the ones that match the filter? (line 137)

Also, rather than completely removing the parkings from the All Spots, can't you keep them but just add a pinned icon to let the user know it is pinned? The pinned icon is not even necessary, but having the parking disappear before your eyes when you pin it defeats the accessibility principle in adding pins in the first place.

File: app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt
Comment (AlexandreBeeDeeBop) [reply app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt:325]: womp womp

**Comment by AlexandreBeeDeeBop** [comment]
> > I dont know about you guys but I HATE the new colours of the screen. The white background on white cards it just awful and now it's even harder to tell if we're in the pinned list or the normal list. Should I change it back to the sort of blue I had before the rebase ?
> 
> I think white cards on white background are alright. What we can do (but I think this is more of a concern for a future sprint) is remove lateral paddings and remove the rounded edges

Weird taste but ok. And what about the little CCTV tick box which is literally invisible now if not activated ? Since it's white box on white background.

**Comment by Armetiis** [comment]
> Weird taste but ok.
Not that weird. If you look on whatsApp or Telegram or any app, you'll see that's basically what they have. Cards that take the width of the whole screen, with non rounded edges and separated by a divider

> And what about the little CCTV tick box which is literally invisible now if not activated ? Since it's white box on white background.
That you broke, because on main it's visible. Make sure to fix that too ^^'

File: app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt
Comment (AlexandreBeeDeeBop) [reply app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt:354]: it could be confusing if the unpin is red and the add favorite is red, what would you want me to do ?

Comment (Armetiis) [reply app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt:354]: You're right. I checked a bit the apps I have. Pin and unpin have the same color, it's the icon that changes.
Also it's not that big of a problem. Suppose we have a paking that's already in favorites. Both swipes are the same color (gray).

You are free to do what you think is right. In my opinion, it's not that bad that they are both reddish

File: app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt
Comment (AlexandreBeeDeeBop) [reply app/src/main/java/com/github/se/cyrcle/ui/list/ListScreen.kt:269]: Disliked it. If you want go on my branch add this line just above the main card and see what it does, it's pretty ugly.

**Review by Armetiis — APPROVED** [review]
Great job! Thanks for the changes

**Review by LouisJamesVasseur — APPROVED** [review]

**Review by Armetiis — APPROVED** [review]

**Comment by AlexandreBeeDeeBop** [comment]
"Review required
Waiting on 1 reapproval from someone other than the last pusher. Reviews from Armetiis and LouisJamesVasseur are stale because they were submitted before the merge base changed"

Please anyone just aprove it's glitched.

**Comment by andrepeiry** [comment]
Merge with admin bypass (GitHub seems to be having a small delay for the check and we don't want to wait). CI is good and has the two approvals of @LouisJamesVasseur and @Armetiis .

**Pull request merged**
